pkg/server/backgroundprofiler/profiler/profiler.go

```diff
@@ -0,0 +1,342 @@
+		startProfile:   make(chan struct{}),
+		startedProfile: make(chan struct{}),
+		stopProfile:    make(chan struct{}),
+		res:            make(chan protoutil.Message),
+	}
+	et.mu.subscribers = make(map[backgroundprofiler.SubscriberID]struct{})
+	et.mu.profileID = 0
+	return et
+}
+
+// CompileTagFilter compiles a filter from the passed in tag. This method is a
+// stripped down version of `compileTagFilter` in
+// https://github.com/google/pprof/blob/main/internal/driver/driver_focus.go#L80.
+func CompileTagFilter(tag string) (func(*profile.Sample) bool, error) {
+	if tag == "" {
+		return nil, nil
+	}
+
+	tagValuePair := strings.SplitN(tag, "=", 2)
+	var wantKey string
+	if len(tagValuePair) == 2 {
+		wantKey = tagValuePair[0]
+		tag = tagValuePair[1]
+	}
+
+	var rfx []*regexp.Regexp
+	for _, tagf := range strings.Split(tag, ",") {
+		fx, err := regexp.Compile(tagf)
+		if err != nil {
+			return nil, errors.Wrap(err, "parsing regexp")
+		}
+		rfx = append(rfx, fx)
+	}
+	if wantKey == "" {
+		return func(s *profile.Sample) bool {
+		matchedrx:
+			for _, rx := range rfx {
+				for key, vals := range s.Label {
+					for _, val := range vals {
+						if rx.MatchString(key + ":" + val) {
+							continue matchedrx
+						}
+					}
+				}
+				return false
+			}
+			return true
+		}, nil
+	}
+	return func(s *profile.Sample) bool {
+		if vals, ok := s.Label[wantKey]; ok {
+			for _, rx := range rfx {
+				for _, val := range vals {
+					if rx.MatchString(val) {
+						return true
+					}
+				}
+			}
+		}
+		return false
+	}, nil
+}
+
+// filterCPUProfileByLabel removes all samples from the profile, except those
+// that match the label regular expression. The filtered profile is written to
+// the passed in profileBytesBuf.
+func filterCPUProfileByLabel(label string, profileBytesBuf *bytes.Buffer) error {
+	p, err := profile.Parse(profileBytesBuf)
+	if err != nil {
+		return err
+	}
+
+	tm, err := CompileTagFilter(label)
+	if err != nil {
+		return err
+	}
+	fm, _ := p.FilterSamplesByTag(tm, nil)
+	if !fm {
+		return nil
+	}
+
+	profileBytesBuf.Reset()
+	return p.Write(profileBytesBuf)
+}
+
+// Start starts the BackgroundProfiler that listens for new Subscribers.
+func (r *BackgroundProfiler) Start(ctx context.Context) error {
+	return r.stopper.RunAsyncTaskEx(ctx, stop.TaskOpts{
+		TaskName: "backgroundprofiler",
+		SpanOpt:  stop.ChildSpan,
+	}, func(ctx context.Context) {
+		ctxDone := ctx.Done()
+		for {
+			select {
+			case <-r.stopper.ShouldQuiesce():
+				return
+			case <-ctxDone:
+				return
+			case <-r.startProfile:
+
+				// Start collecting a CPU profile.
+				var cpuProfileBuf bytes.Buffer
+
+				// TODO(during review): This is an arbitrarily chosen frequency, the
+				// runtime default is a 100Hz. This value will probably require tuning
+				// depending on our use case. Egs: profiling a single query execution vs
+				// profiling a long running job for a few minutes vs profiling a
+				// statement fingerprint.
```

Comment:
We have some sugar around this that we should use

https://github.com/cockroachdb/cockroach/blob/1af66356e77f516ede7cd56a3d7946873a590ef7/pkg/server/status_local_file_retrieval.go#L41-L57